### ISSUE TITLE
QUESTION: will appending data and modifying existing model give better results or faster results?

### ISSUE THREAD

**Issue by sambitmishra98** [pytorch/botorch#951]
I am running the code below to optimise a numerical simulation which takes in 5 columns of inputs (X) and output wall-time (Y) and wish to get the next candidate. At the moment, I am simply running the simulation and using the function below in a loop. 

```
import torch
import numpy as np

from torch                  import tensor, transpose
from botorch.fit            import fit_gpytorch_model as fit_model
from botorch.models         import SingleTaskGP
from botorch.utils          import standardize
from gpytorch.mlls          import ExactMarginalLogLikelihood as EMLL
from botorch.acquisition    import qKnowledgeGradient as qKG
from botorch.optim.optimize import optimize_acqf

def optimise(dat, bounds):

    print("Bayesian Optimisation starts")    
    X = tensor(           list(dat.keys()),    dtype=torch.float32)
    Y = transpose(tensor([list(dat.values())], dtype=torch.float32), 0, 1)
    Y = standardize(Y)

    model = SingleTaskGP(X, -Y)
    fit_model(EMLL(model.likelihood, model))
    candidates, _ = optimize_acqf(
        acq_function = qKG(model, num_fantasies=128), 
        bounds       = tensor(bounds,dtype=torch.float32),
        q            = 1,
        num_restarts = 10,
        raw_samples  = 512,
    )
    print("Bayesian Optimisation ends with next candidate: " , candidates)    

    return list( np.array(candidates[0, :]) )

```

Clearly, I am defining the model and fitting the model with the data in every iteration. This is taking too long. If I try to incrementally add a new value to the existing model (created in the previous iteration say), then will I get the optimisation process to perform faster? Will I get better candidates if I do this? Also, is there such a way to do this?

**Comment by Balandat** [comment]
> Clearly, I am defining the model and fitting the model with the data in every iteration

Just to clarify (given your next question), what is contained in `dat`? This data set should be all previously evaluated candidates (and any other evaluations that you may have sitting around that didn't come from BayesOpt). 

> This is taking too long.

Do you mean model fitting, model fitting and acquisition function optimization, or the overall Bayesian Optimization loop?

> If I try to incrementally add a new value to the existing model (created in the previous iteration say), then will I get the optimisation process to perform faster?

Doing so should help speed up the model fitting process, but it won't speed up the acquisition function optimization process, which usually takes quite a bit longer than fitting. How many data points are you using here? How long does running the simulation take? 

> Will I get better candidates if I do this?

If I understand your question correctly as one of updating the partially fitted model with new data rather than refitting from scratch with all data, then no, this should not do anything to the optimization quality.

> Also, is there such a way to do this?

Yeah what you could do is to use the underlying gpytorch model's `set_train_data` function and then fit the model - the model should retain the optimal hyperparameters from the previous iteration, so calling `fit_model` should be more of a local fine-tuning of the hyperparameters. However, you'd have to be a bit careful about making sure the dimensions are correct (`SingleTaskGP` does a little bit of reshaping magic under the hood). I'd like to understand the specific challenge here first before going down that road. 

Finally, I would recommend using `torch.double` (64 bit instead of 32), since inference for GPs can be tough in lower precision and cause a bunch of numerical issues.

**Comment by sambitmishra98** [comment]
> what is contained in dat?
Yes, `dat` is a dictionary which points input data X to output data Y. I have initialised this as `dat[(val1, val2, val3, val4, val5)] = wall_time`.

> Do you mean model fitting, model fitting and acquisition function optimization, or the overall Bayesian Optimization loop?
I mean the entire process that I am doing in the function above. I want the time taken for the function above to run more efficiently (lesser execution time, better candidates or both).

> How many data points are you using here? How long does running the simulation take?
I am starting with 1 datapoint in the start. I run the simulation with BO and run simulations based on the candidates that the above function suggests. This way I get more data. The simulation takes about 30 minutes, and the function execution time slowly increases (as the data increases) from a few minutes to as much as 30 minutes. 

> If I understand your question correctly as one of updating the partially fitted model with new data rather than refitting from scratch with all data, then no, this should not do anything to the optimization quality.
Thanks for this answer, I always wondered if there is some 'memory' idea that BO uses to suggest the next candidate based on the last candidate (and others before it). I guess BO just fits a model everytime independant of the order of previously suggested candidates and suggests a new one.

> I would recommend using torch.double
Thank you for this suggestion. I will implement this. 

> the model should retain the optimal hyperparameters from the previous iteration

I guess this is exactly what I was looking for. I wish to initialise the model once, and keep adding new data so the model tweaks itself accordingly, hoping to spend lesser time (as opposed to initialising and providing data to fit model from scratch in every iteration).

**Comment by Balandat** [comment]
> I guess this is exactly what I was looking for. I wish to initialise the model once, and keep adding new data so the model tweaks itself accordingly, hoping to spend lesser time (as opposed to initialising and providing data to fit model from scratch in every iteration).

That makes sense, but it may not actually be what's taking the bulk of the time here - typically acquisition function optimization takes a lot longer than the model fitting. What's the breakdown in your case in timing between `fit_model` and `optimize_acqf`? My guess is that most of this is actually coming from `optimize_acqf`.

How many data points do you end up with? 

I see you're using `qKG`, which is relatively computationally expensive, especially when using a large number of fantasies: `num_fantasies=128`. Maybe try something like `num_fantasies=16` or so to see what that does to performance. `qKG` is typically also faster when run on a GPU (if you have one available). Alternatively, you can also use a less expensive acquisition function such as `qEI` or `qNEI` (which is probably something to consider if you're in a regime where the time for generating candidates is on the same order as the simulation time).

**Comment by sambitmishra98** [comment]
> By torch.single do you mean torch.float 32 bit data type? For numerical reasons

Yes, I meant `torch.float32`.
